compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1670,8 +1671,51 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 ),
                 Mismatch::Fixed(s) => (s.into(), s.into(), None),
             };
-            match (&terr, expected == found) {
-                (TypeError::Sorts(values), extra) => {
+            let looks_similar = |e: ExpectedFound<Ty<'_>>| {
+                // We're only interested in adts
+                if let (Some(e), Some(f)) = (e.expected.ty_adt_def(), e.found.ty_adt_def()) {
```

Comment:
Why limit this to ADTs? This can also be true for ADTs named after primitives: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=166bd629e2141b0cd76b038fc470c39a

Reply:
Excellent point, thanks.